pkg/envfuncs/provider_funcs.go

```diff
@@ -0,0 +1,272 @@
+func CreateKindCluster(clusterName string) env.Func {
+	return CreateCluster[kind.Cluster](clusterName)
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateClusterWithConfig which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindClusterWithConfig(clusterName, image, configFilePath string) env.Func {
+	return CreateClusterWithConfig[kind.Cluster](clusterName, configFilePath, kind.WithImage(image))
+}
+
+// Deprecated: This handler has been deprecated in favor of DestroyCluster
+func DestroyKindCluster(name string) env.Func {
+	return DestroyCluster(name)
+}
+
+// Deprecated: This handler has been deprecated in favor of ExportClusterLogs
+func ExportKindClusterLogs(name, dest string) env.Func {
+	return ExportClusterLogs(name, dest)
+}
+
+// CreateCluster returns an env.Func that is used to
+// create an E2E provider cluster that is then injected in the context
+// using the name as a key.
+//
+// NOTE: the returned function will update its env config with the
+// kubeconfig file for the config client.
+func CreateCluster[T provider, P providerRef[T]](clusterName string) env.Func {
+	return func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+		var p P = new(T)
+		k := p.SetDefaults().WithName(clusterName)
+		kubecfg, err := k.Create()
+		if err != nil {
+			return ctx, err
+		}
+
+		// update envconfig  with kubeconfig
+		cfg.WithKubeconfigFile(kubecfg)
+
+		// stall, wait for pods initializations
+		if err := waitForControlPlane(p, cfg.Client()); err != nil {
+			return ctx, err
+		}
+
+		// store entire cluster value in ctx for future access using the cluster name
+		return context.WithValue(ctx, clusterNameContextKey(clusterName), k), nil
+	}
+}
+
+// CreateClusterWithConfig returns an env.Func that is used to
+// create a e2e provider cluster that is then injected in the context
+// using the name as a key.
+//
+// NOTE: the returned function will update its env config with the
+// kubeconfig file for the config client.
+func CreateClusterWithConfig[T provider, P providerRef[T]](clusterName, configFilePath string, opts ...support.ClusterOpts) env.Func {
+	return func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+		var p P = new(T)
+		k := p.SetDefaults().WithName(clusterName).WithOpts(opts...)
+		kubecfg, err := k.CreateWithConfig(configFilePath)
+		if err != nil {
+			return ctx, err
+		}
+
+		// update envconfig  with kubeconfig
+		cfg.WithKubeconfigFile(kubecfg)
+
+		// stall, wait for pods initializations
+		if err := waitForControlPlane(p, cfg.Client()); err != nil {
+			return ctx, err
+		}
+
+		// store entire cluster value in ctx for future access using the cluster name
+		return context.WithValue(ctx, clusterNameContextKey(clusterName), k), nil
+	}
+}
+
+func waitForControlPlane(provider support.E2EClusterProvider, client klient.Client) error {
```

Comment:
I applaud this attempt of making this function convenient for the provider writers. However, I think it makes a lot of assumptions about how a provider works and how it waits for the control plane.  Again, I would suggest making it the responsibility of the provider to know how to wait for its own control plane.

Reply:
https://github.com/kubernetes-sigs/e2e-framework/pull/246#discussion_r1232437324

I am going to move this into the provider.